controllers/storagecluster/cephfilesystem.go

```diff
@@ -162,79 +223,39 @@ func (obj *ocsCephFilesystems) ensureCreated(r *StorageClusterReconciler, instan
-	if err == nil {
-		if existingsvg.DeletionTimestamp != nil {
-			r.Log.Info("Unable to restore subvolumegroup because it is marked for deletion.", "subvolumegroup", klog.KRef(filesystemNamespace, existingsvg.Name))
-			return fmt.Errorf("failed to restore subvolumegroup %s because it is marked for deletion", existingsvg.Name)
-		}
-	}
-
-	cephFilesystemSubVolumeGroup := &cephv1.CephFilesystemSubVolumeGroup{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:            svgName,
-			Namespace:       filesystemNamespace,
-			OwnerReferences: ownerReferences,
-		},
-	}
-
-	// Default value of "distributed" option for pinning in the CephFilesystemSubVolumeGroup CR
-	defaultPinningValue := 1
-	mutateFn := func() error {
-		cephFilesystemSubVolumeGroup.Spec = cephv1.CephFilesystemSubVolumeGroupSpec{
-			Name:           defaultSubvolumeGroupName,
-			FilesystemName: filesystemName,
-			Pinning: cephv1.CephFilesystemSubVolumeGroupSpecPinning{
-				Distributed: &defaultPinningValue,
-			},
+func (r *StorageClusterReconciler) ensureCreateCephFilesystemSubVolumeGroups(filesystem *cephv1.CephFilesystem) error {
+	subVolumeGroups := getSubVolumeGroupsForFilesystem(filesystem)
+	for _, subVolumeGroup := range subVolumeGroups {
+		existing := cephv1.CephFilesystemSubVolumeGroup{}
+		err := r.Client.Get(context.TODO(), types.NamespacedName{Name: subVolumeGroup.Name, Namespace: subVolumeGroup.Namespace}, &existing)
+		switch {
+		case err == nil:
+			if existing.DeletionTimestamp != nil {
+				r.Log.Info("Unable to restore CephFileSystemSubVolumeGroup because it is marked for deletion.", "CephFileSystemSubVolumeGroup", klog.KRef(existing.Namespace, existing.Name))
+				return fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existing.Name)
+			}
+			if !reflect.DeepEqual(existing.Spec, subVolumeGroup.Spec) || !reflect.DeepEqual(existing.ObjectMeta.OwnerReferences, subVolumeGroup.ObjectMeta.OwnerReferences) {
+				r.Log.Info("Restoring original CephFilesystemSubVolumeGroup.", "CephFileSystemSubVolumeGroup", klog.KRef(subVolumeGroup.Namespace, subVolumeGroup.Name))
+				existing.ObjectMeta.OwnerReferences = subVolumeGroup.ObjectMeta.OwnerReferences
+				existing.Spec = subVolumeGroup.Spec
+				err = r.Client.Update(context.TODO(), &existing)
+				if err != nil {
+					r.Log.Error(err, "Unable to update CephFileSystemSubVolumeGroup.", "CephFileSystemSubVolumeGroup", klog.KRef(subVolumeGroup.Namespace, subVolumeGroup.Name))
+					return err
+				}
+			}
+		case errors.IsNotFound(err):
+			r.Log.Info("Creating CephFileSystemSubVolumeGroup.", "CephFileSystemSubVolumeGroup", klog.KRef(subVolumeGroup.Namespace, subVolumeGroup.Name))
+			err = r.Client.Create(context.TODO(), &subVolumeGroup)
+			if err != nil {
+				r.Log.Error(err, "Unable to create CephFileSystemSubVolumeGroup.", "CephFileSystemSubVolumeGroup", klog.KRef(subVolumeGroup.Namespace, subVolumeGroup.Name))
+				return err
+			}
```

Comment:
Why do we create a subvolume group per data pool by default? A sub-volume group requirement is orthogonal to the number of data pools and it is used to segregate access not to designate storage location.

Reply:
@travisn ^^

Reply:
The subvolumegroup per cephfs data pool was requested by the csi team during the design review for this feature. @nixpanic can you remind me why it is necessary? Should the subvolumegroup be per-storageclass or per-datapool?

Reply:
Even i dont see any reason to have SVG per dataPool. lets wait for @nixpanic response as i might have not attended design review.

Reply:
Per discussion today, no subvolumegroup is needed. The storage class that uses the additional data pool just requires the data pool name to be specified as the `pool`